src/template.c

```diff
@@ -7335,6 +7292,83 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
     return nested != 0;
 }
 
+/*****************************************
+ * Append 'this' to the enclosing module members[] so the semantic routines
+ * will get called on the instance members.
+ */
+Dsymbols *TemplateInstance::appendToModuleMember()
+{
+    Module *md = tempdecl->getModule();
+
+    Module *mi = minst;
+    if (mi)
+    {
+        if (mi->isRoot())
+        {
+            /* If both modules mi and md where the template is declared are
+             * roots, instead use md to store the generated code in it.
+             */
+            if (md && md->isRoot())
+                mi = md;
+        }
+        else
+        {
+            /* A module can have explicit template instance and its alias
+             * in module scope (e,g, `alias Base64Impl!('+', '/') Base64;`).
+             * When the module is just imported, normally compiler can assume that
+             * its instantiated code would be contained in the separately compiled
+             * obj/lib file (e.g. phobos.lib).
+             */
+            /*
+             * Bugzilla 2644: However, if the template is instantiated in both
+             * modules of root and non-root, compiler should generate its objcode.
+             * Therefore, always conservatively insert this instance to the member of
+             * a root module, then calculate the necessity by TemplateInstance::needsCodegen().
+             */
+        }
+        //printf("\t1: adding to %s\n", mi->toPrettyChars());
+    }
+    else
+    {
+        // 'this' is speculative instance
+        if (md && md->isRoot())
+        {
+            mi = md;
+        }
+        else
+        {
+            // select arbitrary root module
+            Dsymbol *s = enclosing ? enclosing : tempdecl->parent;
+            while (s && !s->isModule())
+                s = s->toParent2();
+            assert(s);
+            mi = (Module *)s;
+            if (!mi->isRoot())
+                mi = mi->importedFrom;
+        }
+        assert(mi->isRoot());
+        //printf("\t2: adding to module %s instead of module %s\n", mi->toPrettyChars(), sc->module->toPrettyChars());
+    }
+
+    Dsymbols *a = mi->members;
+    for (size_t i = 0; 1; i++)
+    {
+        if (i == a->dim)
+        {
+            a->push(this);
+            if (mi->semanticRun >= PASSsemantic3done && mi->isRoot())
+                Module::addDeferredSemantic3(this);
+            break;
+        }
+        if (this == (*a)[i])    // if already in Array
```

Comment:
I also think the conflict would not happen, but I'm really not sure it yet, because same check existed in old code. So I'd just keep it until we can sure it is unnecessary.